DIPs/DIP1009.md

```diff
@@ -67,34 +94,77 @@ class C {
+
+While this works in the case of non-return identifiers,  proposed solutions must still accomodate declaring the return identifier. All such methods of declaring return identifiers have the advantage of parsing unambiguously with the existing syntax. Thus, simply by omitting the return identifier, they can be used unambiguously to check other variables, which is why the example in the description above used this method. The examples which follow show both the unambiguous way for each method, and the ambiguous way described in 1 above, for comparison.
+
+2. The semicolon style has already been introduced, and is tentatively recommended because:
+* it is the most concise of the alternatives, requiring only `;` to separate the return identifier
+* the semantics of the new `out` expression at least slightly resemble the semantics of `foreach` loops
+```d
+int fun(ref int i)
+  out(r; r > 0)
+  out(; i > 0) // unambiguous parsing
+  out(i > 0) // ambiguous alternative, but possibly manageable (see 1 above)
+{
+  ...
+}
+```
+Note that existing `foreach` statements do not allow omitting the initial identifier, whereas the proposed `out` syntax does allow this. If the ambiguous parsing alternative is rejected, and total consistency is desired, one can either force both to declare an identifier that isn't used, or change `foreach` to allow omitting the identifier.
+
+3. Another possibility is to use two sets of parentheses, which would resemble the existing `out` contract syntax, but also template function declarations. In the absence of a return identifier, one could either require empty parentheses or allow omitting them, in accordance with the restrictions of solution 1 above:
 ```d
 int fun(ref int i)
   out(r)(r > 0)
-  out()(i > 0)
+  out()(i > 0) // unambiguous parsing 
+  out(i > 0) // ambiguous alternative
 {
   ...
 }
 ```
-However, the semicolon syntax is recommended as the "least imperfect" option, because:
-* it resembles `for` and `foreach` loops, and the new `out` expression resembles `foreach` semantics slightly more than it resembles any other semantics
-* it requires only one set of parentheses, which is consistent with the other new contracts
-* it is the most concise of all the proposals, requiring only an additional `;` to achieve clear parsing
+4. Finally, use the lambda syntax in place of the ordinary `assert` condition, e.g.:
+```d
+int fun(ref int i)
+  out(r => r > 0, "blah")
+  out(() => i > 0, "blah") // unambiguous
+  out(i > 0) // ambiguous alternative
+{
+  ...
+}
+```
+The advantage of this is that lambda expression semantics most closely resemble `out` contract semantics. The disadvantage of this variation is that the unambiguous parsing version (`out(() => i > 0, "blah")`), if deemed necessary (see 1 above), is very awkward and verbose.
```

Comment:
> Why not just use out(_ => i > 0)?

That is an option available to all the syntaxes. This particular instance is not attractive to me at all. Such a weird looking thing to have to create an unused variable AND a lambda expression merely to check an existing variable. That said, almost all of the available options are serviceable, meaning that whatever is chosen, people will likely get used to them.